### DISCUSSION TITLE
Changing SympifyError in __mul__ and __rmul__ to NotImplemented

### DISCUSSION THREAD

**Discussion by Kadelka** [sympy/sympy#26240] [General]
My code for the class Octonion in modul octonion.py [octonion.zip](https://github.com/sympy/sympy/files/14303275/octonion.zip) uses class Quaternion in sympy. If `a` is an instance of Octonion and `b` an instance of Quaternion, then `a*b` is correctly evaluated as `a.__mul__(b)`. But `b*a` doesn't do what it should do: I get `SympifyError` since `b*a` is resolved as `b.__mul__(a)` and Quaternions don't know Octonions.

Is there any idea to solve this problem? Of course I don't want change sympy. Is it possible that for an Octonion `a` `Quaternion.__mul__(self,a)` returns a NotImplemented instead of stopping with a SympifyError?

**Comment by oscarbenjamin** [comment]
NotImplemented should be returned

**Reply by Kadelka** [reply]
Is it a bug in sympy.Quaternion or in my code? See the output in my additional comment.

**Reply by Kadelka** [reply]
Even `'s'*a` triggers the same phenomenon.

**Reply by Kadelka** [reply]
and even 's'*Quaternion(1,2,3,4).

**Comment by Kadelka** [comment]
Unfortunately NotImplemented is not returned. Here is what I get:
```Python 3.11.8 (main, Feb  8 2024, 09:59:21) [GCC 13.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from octonion import *
>>> (4+5j)*a
octonion(4.0*a0 - 5.0*a1, 5.0*a0 + 4.0*a1, 4.0*a2 - 5.0*a3, 5.0*a2 + 4.0*a3, 4.0*a4 - 5.0*a5, 5.0*a4 + 4.0*a5, 4.0*a6 + 5.0*a7, -5.0*a6 + 4.0*a7)
>>> Quaternion(1,2,3)*a
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.11/site-packages/sympy/algebras/quaternion.py", line 680, in __mul__
    return self._generic_mul(self, _sympify(other))
                                   ^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/sympy/core/sympify.py", line 528, in _sympify
    return sympify(a, strict=True)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/sympy/core/sympify.py", line 449, in sympify
    raise SympifyError(a)
sympy.core.sympify.SympifyError: SympifyError: octonion(a0, a1, a2, a3, a4, a5, a6, a7)
>```
